Library/Homebrew/cli_parser.rb

```diff
@@ -0,0 +1,280 @@
+        end
+      end
+
+      required.each do |sym|
+        raise StandardError, "option --#{@specs[sym].long} must be specified" unless given_args.include? sym
+      end
+
+      given_args.each do |sym, given_data|
+        arg, params = given_data.values_at :arg, :params
+
+        opts = @specs[sym]
+        if params.empty? && !opts.flag?
+          raise StandardError, "option #{arg} needs a parameter" unless opts.default
+          params << [opts.default]
+        end
+
+        case opts.type
+        when :string, :strings
+          vals[sym] = params.map { |pg| pg.map(&:to_s) }
+        when :flag
+          vals[sym] = true
+        end
+
+        if opts.single_arg?
+          if opts.multi?
+            vals[sym] = vals[sym].map { |p| p[0] }
+          else
+            vals[sym] = vals[sym][0][0]
+          end
+        elsif opts.multi_arg? && !opts.multi?
+          vals[sym] = vals[sym][0]
+        end
+      end
+
+      cmdline.clear
+      @unnamed.each { |l| cmdline << l }
+
+      class << vals
+        def method_missing(m, *_args)
+          if key?(m) || key?(m.to_s)
+            self[m] || self[m.to_s]
+          else
+            super
+          end
+        end
+
+        # https://robots.thoughtbot.com/always-define-respond-to-missing-when-overriding
+        def respond_to_missing?(m, include_private = false)
+          super
+        end
+      end
+      vals
+    end
+
+    private
+
+    def each_arg(args)
+      remains = []
+      i = 0
+
+      until i >= args.length
+        # return remains += args[i..-1] if @stop_words.member? args[i]
+        case args[i]
+        when /^--$/
+          return remains += args[i..-1]
+        when /^--(\S+?)=(.*)$/
+          num_params_taken = yield "--#{Regexp.last_match(1)}", [Regexp.last_match(2)]
+          remains << args[i] if num_params_taken.nil?
+          i += 1
+        when /^--(\S+)$/
+          params = collect_argument_parameters(args, i + 1)
+          num_params_taken = yield args[i], params
+          if num_params_taken.nil?
+            remains << args[i]
+          else
+            i += num_params_taken
+          end
+          i += 1
+        else
+          remains << args[i]
+          i += 1
+        end
+      end
+
+      remains
+    end
+
+    def collect_argument_parameters(args, start_at)
+      params = []
+      pos = start_at
+      while args[pos] && args[pos] !~ PARAM_RE
+        params << args[pos]
+        pos += 1
+      end
+      params
+    end
+  end
+
+  PARAM_RE = /^-(-|\.$|[^\d\.])/
+
+  class Option
+    attr_accessor :name, :opts
+
+    FLAG_TYPES = [:flag, :bool, :boolean].freeze
+
+    SINGLE_ARG_TYPES = [:string].freeze
+
+    MULTI_ARG_TYPES = [:strings].freeze
+
+    def initialize(name, _desc = "", opts = {})
+      opts[:type] = case opts[:type]
+      when :boolean, :bool then :flag
+      when :string then :string
+      when Class
```

Comment:
What's this used for currently?

Reply:
This can be used for may be datetime class , I’ve seen some date arguments in Homebrew . Anyways I think I’m trying to do too much in this PR supporting named args , switches and named args with multiple params. I’m stripping all the other features amd rewriting with implementation of named and boolean arguments

Reply:
I think that sounds very sensible @GauthamGoli.